File: deepCopy.java
//Time Complexity: O(n) length of list
//Space Complexity: O(n) for the hashamp
//Did it run on leetcode: yes
/** Approach: First I have iterated the linked list and created new nodes and have put them in hashmap
hashmap contains original node as key and new node as value
Second I have connected the new nodes with taking new head
Third using the hashmap I have assigned random pointers 
Since I used a hashmap to get the reference takes only O(1) time
**/

class Solution {
    public Node copyRandomList(Node head) {
        if(head==null)
            return null;
        
        Map<Node, Node> map = new HashMap<>();
        
        Node dummy1 = head;
        
        while(dummy1!=null){
            Node newNode = new Node(dummy1.val);
            map.put(dummy1,newNode);
            dummy1 = dummy1.next;
            newNode = newNode.next;
        }
        dummy1 = head;
       
        Node newHead = map.get(head);
        Node dummy2 = newHead;
        while(dummy1.next!=null){
            dummy2.next = map.get(dummy1.next);
            dummy1 = dummy1.next;
            dummy2= dummy2.next;
        }
        
        dummy1 = head;
        dummy2 = newHead;
        
        while(dummy1!=null){
            if(map.containsKey(dummy1.random))
                dummy2.random = map.get(dummy1.random);
            else
                dummy2.random = null;
            dummy1 = dummy1.next;
            dummy2 = dummy2.next;
        }
        
        return newHead;
    }
}

File: deepcopy.py
#Time Complexity: O(n)
#Space Complexity: O(1)
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        
        if head == None:
            return None
        curr = head
        while curr:
            newnode = Node(curr.val)
            newnode.next = curr.next
            curr.next = newnode
            curr = curr.next.next
            
        curr = head
        
        while curr:
            if curr.random != None:
                curr.next.random = curr.random.next
            curr = curr.next.next
            
        curr = head
        deepcopy = head.next
        copycurr = deepcopy
        
        while curr.next.next:
            curr.next = curr.next.next
            deepcopy.next = deepcopy.next.next
            curr = curr.next
            deepcopy  = deepcopy.next
            
            
        return copycurr
            
        
            
            